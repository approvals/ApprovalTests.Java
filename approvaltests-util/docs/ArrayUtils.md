<a id="top"></a>

# ArrayUtils

<!-- toc -->
## Contents

  * [toArray](#toarray)<!-- endToc -->

## toArray

ArrayUtils has the ability to dynamically determine the type of your list.

<!-- snippet: toArray -->
<a id='snippet-toArray'></a>
```java
List<Comparable> comparables = new ArrayList<>();
comparables.add(null);
comparables.add(1);
comparables.add(3.1415);
comparables.add("Lars");
Comparable[] comparableArray = ArrayUtils.toArray(comparables);
```
<sup><a href='/approvaltests-util-tests/src/test/java/com/spun/util/ArrayUtilsTest.java#L105-L112' title='Snippet source file'>snippet source</a> | <a href='#snippet-toArray' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

This works in almost all cases with the notable exceptions of empty arrays, some common interfaces 
which will not be preferenced.
In which case you can always use the overloaded method 
<!-- snippet: toArrayWithClass -->
<a id='snippet-toArrayWithClass'></a>
```java
Comparable[] array = ArrayUtils.toArray(comparables, Comparable.class);
```
<sup><a href='/approvaltests-util-tests/src/test/java/com/spun/util/ArrayUtilsTest.java#L113-L115' title='Snippet source file'>snippet source</a> | <a href='#snippet-toArrayWithClass' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->


[Back to User Guide](README.md#top)
