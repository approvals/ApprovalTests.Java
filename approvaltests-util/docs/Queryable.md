<!--
GENERATED FILE - DO NOT EDIT
This file was generated by [MarkdownSnippets](https://github.com/SimonCropp/MarkdownSnippets).
Source File: /approvaltests-util/docs/mdsource/Queryable.source.md
To change this file edit the source file and then run MarkdownSnippets.
-->

<a id="top"></a>

# Queryable

<!-- toc -->
## Contents

  * [Extending Queryable](#extending-queryable)<!-- endToc -->

## Extending Queryable

You have a function you wish was part of Queryable.
<!-- snippet: custom-query -->
<a id='snippet-custom-query'></a>
```java
public static Queryable<String> findFirstWordsOnly(List<String> words)
{
  return Query.select(words, w -> {
    int i = w.indexOf(' ');
    if (i == -1)
    {
      return w;
    }
    else
    {
      return w.substring(0, i);
    }
  });
}
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.java#L30-L45' title='File snippet `custom-query` was extracted from'>snippet source</a> | <a href='#snippet-custom-query' title='Navigate to start of snippet `custom-query`'>anchor</a></sup>
<!-- endSnippet -->
You can add this to Queryable by implementing the `com.lambda.utils.Extendable` interface.
<!-- snippet: implementing-extendable -->
<a id='snippet-implementing-extendable'></a>
```java
public static class CustomQuery implements Extendable<List<String>>
{
  private List<String> caller;

  @Override public void setCaller(List<String> caller) {
    this.caller = caller;
  }
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.java#L14-L22' title='File snippet `implementing-extendable` was extracted from'>snippet source</a> | <a href='#snippet-implementing-extendable' title='Navigate to start of snippet `implementing-extendable`'>anchor</a></sup>
<!-- endSnippet -->
Now you can add extension methods that are **not static**
<!-- snippet: extendable-query -->
<a id='snippet-extendable-query'></a>
```java
public Queryable<String> findFirstWordsOnly() {
  return findFirstWordsOnly(caller);
}
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.java#L24-L28' title='File snippet `extendable-query` was extracted from'>snippet source</a> | <a href='#snippet-extendable-query' title='Navigate to start of snippet `extendable-query`'>anchor</a></sup>
<!-- endSnippet -->
and now you can call it as such
<!-- snippet: custom-query-example -->
<a id='snippet-custom-query-example'></a>
```java
Queryable<String> list = Queryable.as("One fish", "two fish", "red fish", "blue fish");
Queryable<String> firstWordsOnlyWithExtension = list.select(String::toUpperCase).use(CustomQuery.class).findFirstWordsOnly();
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.java#L50-L53' title='File snippet `custom-query-example` was extracted from'>snippet source</a> | <a href='#snippet-custom-query-example' title='Navigate to start of snippet `custom-query-example`'>anchor</a></sup>
<!-- endSnippet -->
whereas previously you had to use
<!-- snippet: custom-query-example-static -->
<a id='snippet-custom-query-example-static'></a>
```java
Queryable<String> firstWordsOnlyStatic = CustomQuery.findFirstWordsOnly(Query.select(list, String::toUpperCase));
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.java#L54-L56' title='File snippet `custom-query-example-static` was extracted from'>snippet source</a> | <a href='#snippet-custom-query-example-static' title='Navigate to start of snippet `custom-query-example-static`'>anchor</a></sup>
<!-- endSnippet -->

---

[Back to User Guide](README.md#top)
