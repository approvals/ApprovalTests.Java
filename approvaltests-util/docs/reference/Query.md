<a id="top"></a>

# Query and Queryable

<!-- toc -->
## Contents

  * [What it is](#what-it-is)
    * [Comparison vs. Streams](#comparison-vs-streams)
      * [Other benefits](#other-benefits)
    * [Queryable extends List](#queryable-extends-list)
    * [API](#api)
      * [SelectMany](#selectmany)
        * [SelectMany vs SelectManyArray](#selectmany-vs-selectmanyarray)
        * [Example:](#example)
    * [GroupBy](#groupby)
      * [Full transforms](#full-transforms)
  * [HowTos](#howtos)<!-- endToc -->

## What it is
Query (and the Queryable wrapper) is an alternative to Streams. In other words it is an implementation of Map/Reduce 
based on the naming style of MS Linq.

### Comparison vs. Streams
Here is an example of filtering a list of numbers for only even numbers, sorted in reverse order:

<!-- snippet: query_example -->
<a id='snippet-query_example'></a>
```java
Integer[] numbers = Range.get(1, 20);
Integer[] evenQueryNumbers = Query.where(numbers, n -> n % 2 == 0).orderBy(OrderBy.Order.Descending, n -> n)
    .asArray();
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryTest.java#L37-L41' title='Snippet source file'>snippet source</a> | <a href='#snippet-query_example' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

Here is the exact same function but using Java Streams:

<!-- snippet: stream_example -->
<a id='snippet-stream_example'></a>
```java
Integer[] evenStreamNumbers = Arrays.stream(numbers).filter(n -> n % 2 == 0)
    .sorted((o1, o2) -> o2.compareTo(o1)).toArray(Integer[]::new);
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryTest.java#L42-L45' title='Snippet source file'>snippet source</a> | <a href='#snippet-stream_example' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

Another example using sum() vs. reduce():

<!-- snippet: query_sum_example -->
<a id='snippet-query_sum_example'></a>
```java
String[] names = {"Llewellyn", "Scott"};
int lengthsFromQuery = Query.sum(names, n -> n.length()).intValue();
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryTest.java#L49-L52' title='Snippet source file'>snippet source</a> | <a href='#snippet-query_sum_example' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

Here is the exact same function but using Java Streams:

<!-- snippet: stream_sum_example -->
<a id='snippet-stream_sum_example'></a>
```java
int lengthsFromStream = (int) Arrays.stream(names).map(n -> n.length()).reduce(0, (a, b) -> a + b);
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryTest.java#L53-L55' title='Snippet source file'>snippet source</a> | <a href='#snippet-stream_sum_example' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

#### Other benefits
* Simpler interface
* Queryable<T> extends List<T>
* Queryable is customizable
* Uses SQL syntax

### Queryable extends List
Another advantage of Query and Queryable is that Queryable extends List. Here is an example of converting
an Array of Integers to Strings:

Using Query:
<!-- snippet: list_is_queryable -->
<a id='snippet-list_is_queryable'></a>
```java
List<String> strings = Query.select(numbers, n -> "" + n);
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryTest.java#L60-L62' title='Snippet source file'>snippet source</a> | <a href='#snippet-list_is_queryable' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

Using Streams:
<!-- snippet: list_from_stream -->
<a id='snippet-list_from_stream'></a>
```java
List<String> strings2 = Arrays.stream(numbers).map(n -> "" + n).collect(Collectors.toList());
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryTest.java#L63-L65' title='Snippet source file'>snippet source</a> | <a href='#snippet-list_from_stream' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

### API

#### SelectMany

SelectMany allows you to flatten query results.

For example:  

If each element of your SELECT statement is a List<String> you would get back a List<List<String>>. If you would like to flatten that into a List<String>, the easiest way is to do a Query.selectMany(). 

##### SelectMany vs SelectManyArray

Because of the limitations of the Java language, we could not override the function to allow you to return either a list or an array in your query. As a workaround we overrode the name, so if you are returning a String[] you would need to call selectManyArray.

##### Example:

The following code extracts the words from the following sentences, and orders them alphabetically:

<!-- snippet: queryable_select_many -->
<a id='snippet-queryable_select_many'></a>
```java
Queryable<String> names = Queryable.as("Now is the time", "Fourscore and seven years ago",
    "When in the course of human events");
Queryable<String> allNames = names.selectMany(n -> Arrays.asList(n.split(" "))).orderBy(n -> n);
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.java#L142-L146' title='Snippet source file'>snippet source</a> | <a href='#snippet-queryable_select_many' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

resulting in

<!-- snippet: QueryableTest.testSelectManyCharacters.approved.txt -->
<a id='snippet-QueryableTest.testSelectManyCharacters.approved.txt'></a>
```txt
[0] = Fourscore
[1] = Now
[2] = When
[3] = ago
[4] = and
[5] = course
[6] = events
[7] = human
[8] = in
[9] = is
[10] = of
[11] = seven
[12] = the
[13] = the
[14] = time
[15] = years
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.testSelectManyCharacters.approved.txt#L1-L16' title='Snippet source file'>snippet source</a> | <a href='#snippet-QueryableTest.testSelectManyCharacters.approved.txt' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->


### GroupBy

Here is a simple example of grouping words by their first letter.
<!-- snippet: group_by_key -->
<a id='snippet-group_by_key'></a>
```java
Queryable<String> words = Queryable.as("Jack", "and", "Jill", "jumped", "up", "the", "hill");
Queryable<Entry<Character, Queryable<String>>> result = words.groupBy(w -> w.toLowerCase().charAt(0));
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.java#L152-L155' title='Snippet source file'>snippet source</a> | <a href='#snippet-group_by_key' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->
producing:
<!-- snippet: QueryableTest.testGroupBy.approved.txt -->
<a id='snippet-QueryableTest.testGroupBy.approved.txt'></a>
```txt
[0] = j=[Jack, Jill, jumped]
[1] = a=[and]
[2] = u=[up]
[3] = t=[the]
[4] = h=[hill]
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.testGroupBy.approved.txt#L1-L5' title='Snippet source file'>snippet source</a> | <a href='#snippet-QueryableTest.testGroupBy.approved.txt' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

#### Full transforms

In a group by, you can select both how the key is selected and how each object for the key is selected,
as well as how the final list is transformed.

Here is an example of doing all three to count the number of words of the same length:
<!-- snippet: group_by_full -->
<a id='snippet-group_by_full'></a>
```java
Queryable<String> words = Queryable.as("One Fish Two Fish Red Fish Blue Fish".split(" "));
Queryable<Entry<Object, Object>> result = words.groupBy(w -> w.length(), w -> w.toLowerCase(),
    r -> r.join("_"));
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.java#L175-L179' title='Snippet source file'>snippet source</a> | <a href='#snippet-group_by_full' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->
resulting in
<!-- snippet: QueryableTest.testGroupByCombineWordsOfSimilarLengths.approved.txt -->
<a id='snippet-QueryableTest.testGroupByCombineWordsOfSimilarLengths.approved.txt'></a>
```txt
3 = one_two_red
4 = fish_fish_fish_blue_fish
```
<sup><a href='/approvaltests-util-tests/src/test/java/org/lambda/query/QueryableTest.testGroupByCombineWordsOfSimilarLengths.approved.txt#L1-L2' title='Snippet source file'>snippet source</a> | <a href='#snippet-QueryableTest.testGroupByCombineWordsOfSimilarLengths.approved.txt' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

## HowTos
* [How to extend Queryable](../how_to/ExtendQueryable.md#top)


---

[Back to User Guide](README.md#top)
