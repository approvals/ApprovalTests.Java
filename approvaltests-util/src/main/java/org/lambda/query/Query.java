package org.lambda.query;

import java.util.*;

import org.lambda.functions.Function1;
import org.lambda.query.OrderBy.Order;

import com.spun.util.ArrayUtils;

public class Query<In>
{
  public static <In, Out> Queryable<Out> select(Collection<In> list, Function1<In, Out> function)
  {
    Queryable<Out> out = new Queryable<Out>();
    for (In i : list)
    {
      out.add(function.call(i));
    }
    return out;
  }
  public static <In, Out> Queryable<Out> select(In[] list, Function1<In, Out> function)
  {
    return select(ArrayUtils.asList(list), function);
  }
  public static <In> Queryable<In> where(Iterable<In> list, Function1<In, Boolean> funct)
  {
    Queryable<In> out = new Queryable<In>();
    for (In i : list)
    {
      if (funct.call(i))
      {
        out.add(i);
      }
    }
    return out;
  }
  public static <In> In first(In[] list, Function1<In, Boolean> filter)
  {
    return first(ArrayUtils.asList(list), filter);
  }
  public static <In> In first(Iterable<In> list, Function1<In, Boolean> filter)
  {
    for (In i : list)
    {
      if (filter.call(i))
      { return i; }
    }
    return null;
  }
  public static <In> Queryable<In> where(In[] list, Function1<In, Boolean> filter)
  {
    Queryable<In> out = new Queryable<In>();
    for (In i : list)
    {
      if (filter.call(i))
      {
        out.add(i);
      }
    }
    return out;
  }
  public static <In, Out extends Comparable<Out>> In max(Collection<In> list, Function1<In, Out> f1)
  {
    return getTop(list, f1, 1);
  }
  public static <In, Out extends Comparable<Out>> In min(List<In> list, Function1<In, Out> f1)
  {
    return getTop(list, f1, -1);
  }
  public static <In> Double average(List<In> list, Function1<In, Number> f1)
  {
    double total = 0.00;
    for (In in : list)
    {
      total += f1.call(in).doubleValue();
    }
    return total / list.size();
  }
  private static <In, Out extends Comparable<Out>> In getTop(Collection<In> list, Function1<In, Out> f1,
      int modifier)
  {
    if (ArrayUtils.isEmpty(list))
    { return null; }
    In found = list.iterator().next();
    Out max = f1.call(found);
    for (In in : list)
    {
      Out current = f1.call(in);
      if (max.compareTo(current) * modifier < 0)
      {
        max = current;
        found = in;
      }
    }
    return found;
  }
  public static <T, Out extends Comparable<Out>> T[] orderBy(T[] list, Function1<T, Out> f1)
  {
    return orderBy(list, Order.Ascending, f1);
  }
  public static <T, Out extends Comparable<Out>> T[] orderBy(T[] list, Order order, Function1<T, Out> f1)
  {
    Arrays.sort(list, new OrderBy<T, Out>(order, f1));
    return list;
  }
  public static <T> Queryable<T> orderBy(List<T> list, Function1<T, Comparable<?>> f1)
  {
    return orderBy(list, Order.Ascending, f1);
  }
  public static <T> Queryable<T> orderBy(List<T> list, Order order, Function1<T, Comparable<?>> f1)
  {
    Collections.sort(list, new OrderBy<T, Comparable<?>>(order, f1));
    return Queryable.as(list);
  }
  public static <In, Out extends Number> Double sum(In[] list, Function1<In, Out> f1)
  {
    return sum(ArrayUtils.asList(list), f1);
  }
  public static <In, Out extends Number> Double sum(Collection<In> list, Function1<In, Out> f1)
  {
    double sum = 0;
    for (In in : list)
    {
      sum += f1.call(in).doubleValue();
    }
    return sum;
  }
  public static <Out extends Number> Double sum(Collection<Out> list)
  {
    return sum(list, a -> a);
  }
  public static <Out extends Number> Double sum(Out[] list)
  {
    return sum(list, a -> a);
  }
  @SuppressWarnings({"rawtypes", "unchecked"})
  public static <T extends Number> T max(List<T> numbers)
  {
    return (T) max(numbers, (a) -> (Comparable) a);
  }
  public static <T extends Number> T max(T[] numbers)
  {
    return max(ArrayUtils.asList(numbers));
  }
  @SuppressWarnings({"rawtypes", "unchecked"})
  public static <T extends Number> T min(List<T> numbers)
  {
    return (T) min((List) numbers, (Comparable a) -> a);
  }
  public static <In> boolean all(In[] array, Function1<In, Boolean> funct)
  {
    return array.length == where(array, funct).size();
  }
  public static <In> boolean all(List<In> array, Function1<In, Boolean> funct)
  {
    return array.size() == where(array, funct).size();
  }
  public static <In> boolean any(List<In> array, Function1<In, Boolean> funct)
  {
    return first(array, funct) != null;
  }
  public static <In> boolean any(In[] array, Function1<In, Boolean> funct)
  {
    return first(array, funct) != null;
  }
  public static <In> Queryable<In> distinct(List<In> list)
  {
    Queryable<In> distinct = new Queryable<>();
    for (In in : list)
    {
      if (!distinct.contains(in))
      {
        distinct.add(in);
      }
    }
    return distinct;
  }
  public static <In> In last(In[] list)
  {
    return last(Arrays.asList(list));
  }
  static <In> In last(List<In> asList)
  {
    if (asList.isEmpty())
    { return null; }
    return asList.get(asList.size() - 1);
  }
}
