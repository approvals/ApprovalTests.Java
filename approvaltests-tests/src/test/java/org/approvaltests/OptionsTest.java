package org.approvaltests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.File;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.List;

import org.approvaltests.awt.AwtApprovals;
import org.approvaltests.combinations.CombinationApprovals;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.core.Options;
import org.approvaltests.reporters.FirstWorkingReporter;
import org.approvaltests.reporters.UseReporter;
import org.approvaltests.reporters.UseReporterTest;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.lambda.query.Query;
import org.lambda.query.Queryable;

import com.spun.util.ArrayUtils;

public class OptionsTest
{
  @Test
  void verifyWithReporterInOptions()
  {
    ApprovalFailureReporterSpy reporter = new ApprovalFailureReporterSpy();
    Options options = new Options(reporter);
    try
    {
      Approvals.verify("hallo", options);
    }
    catch (Error error)
    {
      assertTrue(reporter.hasBeenCalled());
    }
  }
  @Test
  void verifyWithReporterFromOptionsBuilder()
  {
    ApprovalFailureReporterSpy reporter = new ApprovalFailureReporterSpy();
    Options options = new Options().withReporter(reporter);
    try
    {
      Approvals.verify("hallo", options);
    }
    catch (Error error)
    {
      assertTrue(reporter.hasBeenCalled());
    }
  }
  @Test
  @UseReporter(UseReporterTest.TestReporter.class)
  void verifyWithoutReporter()
  {
    FirstWorkingReporter reporter = (FirstWorkingReporter) new Options().getReporter();
    assertEquals(UseReporterTest.TestReporter.class, reporter.getReporters()[1].getClass());
  }
  @Test
  void testTheVerifyApi()
  {
    Method[] declaredMethods = Approvals.class.getDeclaredMethods();
    List<Method> methodList = Query.where(declaredMethods, m -> m.getName().startsWith("verify"));
    methodList = Query.orderBy(methodList, m -> m.toString());
    Approvals.verifyAll("verifyMethods", methodList, m -> m.toString(), new Options());
  }
  @Test
  void testEachMethodHasOneWithOptions()
  {
    for (Class c : getApprovalClasses())
    {
      verifyEachVerifyMethodHasOneWithOptions(c);
    }
  }
  public static List<Class<?>> getApprovalClasses()
  {
    return Arrays.asList(Approvals.class, CombinationApprovals.class, AwtApprovals.class);
  }
  private void verifyEachVerifyMethodHasOneWithOptions(Class<?> approvalsClass)
  {
    Queryable<Method> declaredMethods = Queryable.as(approvalsClass.getDeclaredMethods());
    Queryable<Method> methodList = declaredMethods.where(m -> m.getName().startsWith("verify")
        && Modifier.isPublic(m.getModifiers()) && !m.isAnnotationPresent(Deprecated.class));
    List<Method> methodsWithOptions = methodList.where(m -> isOptionsPresent(m));
    List<Method> methodsWithoutOptions = methodList.where(m -> !(isOptionsPresent(m)));
    for (Method withOptions : methodsWithOptions)
    {
      String name = withOptions.getName();
      Class<?>[] parameterTypes = withOptions.getParameterTypes();
      Class<?>[] parameters = getWithoutOptions(parameterTypes);
      boolean foundWithoutOptions = Query.any(methodsWithoutOptions,
          m -> name.equals(m.getName()) && Arrays.deepEquals(m.getParameterTypes(), parameters));
      assertTrue(foundWithoutOptions, "No match found for:" + withOptions);
    }
    for (Method withoutOptions : methodsWithoutOptions)
    {
      String name = withoutOptions.getName();
      Class[] parameters = withoutOptions.getParameterTypes();
      boolean foundWithOptions = Query.any(methodsWithOptions, m -> name.equals(m.getName())
          && Arrays.deepEquals(getWithoutOptions(m.getParameterTypes()), parameters));
      assertTrue(foundWithOptions, "No match found for:\n" + withoutOptions);
    }
    assertEquals(methodsWithOptions.size(), methodsWithoutOptions.size());
    assertNotEquals(0, methodsWithoutOptions.size());
  }
  private Class<?>[] getWithoutOptions(Class<?>[] parameterTypes)
  {
    Class<?>[] parameters = Query.where(parameterTypes, t -> !t.equals(Options.class)).toArray(new Class[0]);
    return parameters;
  }
  public static boolean isOptionsPresent(Method m)
  {
    return ArrayUtils.getLast(m.getParameterTypes()).equals(Options.class)
        || ArrayUtils.getFirst(m.getParameterTypes()).equals(Options.class);
  }
  @Test
  void verifyFileExtension()
  {
    Approvals.verify("<html><body><h1>hello approvals</h1></body></html>",
        new Options().forFile().withExtension(".html"));
  }
  @Test
  void verifyFileName()
  {
    String sampleText = "<html><body><h1>hello approvals</h1></body></html>";
    Approvals.verify(sampleText,
        new Options().forFile().withBaseName("customApproval").forFile().withExtension(".html"));
    Approvals.verify(sampleText, new Options().forFile().withName("customApproval", ".html"));
    Approvals.verify(sampleText, new Options().forFile().withBaseName("customApproval"));
  }
  static class ApprovalFailureReporterSpy implements ApprovalFailureReporter
  {
    private boolean hasBeenCalled;
    public ApprovalFailureReporterSpy()
    {
      this.hasBeenCalled = false;
    }
    @Override
    public void report(String received, String approved)
    {
      hasBeenCalled = true;
    }
    public boolean hasBeenCalled()
    {
      return hasBeenCalled;
    }
  }
  @Test
  void testDefaultWriterForOptions()
  {
    final Options options = new Options();
    ApprovalWriter writer = options.createWriter("any text");
    Approvals.verify(writer);
  }
  @Test
  void testCustomWriter()
  {
    final Options options = new Options();
    ApprovalWriter anotherWriter = options.withWriter(MyWriter::getFactory).createWriter("any text");
    assertTrue(anotherWriter instanceof MyWriter);
  }
  public static class MyWriter implements ApprovalWriter
  {
    public static ApprovalWriter getFactory(Object content, Options options)
    {
      return new MyWriter();
    }
    @Override
    public File writeReceivedFile(File received)
    {
      return null;
    }
    @Override
    public String getFileExtensionWithDot()
    {
      return null;
    }
  }
}
