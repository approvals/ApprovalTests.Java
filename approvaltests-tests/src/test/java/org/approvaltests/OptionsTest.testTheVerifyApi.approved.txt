verifyMethods


private static void org.approvaltests.Approvals.verifyEachFileAgainstMasterDirectory(java.io.File[],org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(com.spun.util.persistence.ExecutableCommand)
public static void org.approvaltests.Approvals.verify(com.spun.util.persistence.ExecutableCommand,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(com.spun.util.persistence.SqlLoader)
public static void org.approvaltests.Approvals.verify(com.spun.util.persistence.SqlLoader,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(java.io.File)
public static void org.approvaltests.Approvals.verify(java.io.File,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(java.lang.Object)
public static void org.approvaltests.Approvals.verify(java.lang.Object,java.lang.String)
public static void org.approvaltests.Approvals.verify(java.lang.Object,java.lang.String,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(java.lang.Object,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(java.lang.String)
public static void org.approvaltests.Approvals.verify(java.lang.String,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(java.sql.ResultSet)
public static void org.approvaltests.Approvals.verify(java.sql.ResultSet,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(java.util.Map)
public static void org.approvaltests.Approvals.verify(java.util.Map,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(org.approvaltests.approvers.ApprovalApprover)
public static void org.approvaltests.Approvals.verify(org.approvaltests.approvers.ApprovalApprover,org.approvaltests.core.ApprovalFailureReporter)
public static void org.approvaltests.Approvals.verify(org.approvaltests.approvers.ApprovalApprover,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(org.approvaltests.core.ApprovalWriter)
public static void org.approvaltests.Approvals.verify(org.approvaltests.core.ApprovalWriter,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(org.approvaltests.core.ApprovalWriter,org.approvaltests.namer.ApprovalNamer)
public static void org.approvaltests.Approvals.verify(org.approvaltests.core.ApprovalWriter,org.approvaltests.namer.ApprovalNamer,org.approvaltests.core.ApprovalFailureReporter)
public static void org.approvaltests.Approvals.verify(org.approvaltests.core.ApprovalWriter,org.approvaltests.namer.ApprovalNamer,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verify(org.approvaltests.core.Verifiable)
public static void org.approvaltests.Approvals.verify(org.approvaltests.core.Verifiable,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAll(java.lang.Iterable,org.lambda.functions.Function1)
public static void org.approvaltests.Approvals.verifyAll(java.lang.Iterable,org.lambda.functions.Function1,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAll(java.lang.Object[],org.lambda.functions.Function1)
public static void org.approvaltests.Approvals.verifyAll(java.lang.Object[],org.lambda.functions.Function1,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.Iterable)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.Iterable,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.Iterable,org.lambda.functions.Function1)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.Iterable,org.lambda.functions.Function1,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.Object[])
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.Object[],org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.Object[],org.lambda.functions.Function1)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.Object[],org.lambda.functions.Function1,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.String,java.lang.Iterable)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.String,java.lang.Iterable,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.String,java.lang.Object[])
public static void org.approvaltests.Approvals.verifyAll(java.lang.String,java.lang.String,java.lang.Object[],org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAsJson(java.lang.Object)
public static void org.approvaltests.Approvals.verifyAsJson(java.lang.Object,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyAsJson(java.lang.Object,org.lambda.functions.Function1,java.lang.Class)
public static void org.approvaltests.Approvals.verifyAsJson(java.lang.Object,org.lambda.functions.Function1,java.lang.Class,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyEachFileInDirectory(java.io.File)
public static void org.approvaltests.Approvals.verifyEachFileInDirectory(java.io.File,java.io.FileFilter)
public static void org.approvaltests.Approvals.verifyEachFileInDirectory(java.io.File,java.io.FileFilter,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyEachFileInDirectory(java.io.File,java.io.FilenameFilter)
public static void org.approvaltests.Approvals.verifyEachFileInDirectory(java.io.File,java.io.FilenameFilter,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyEachFileInDirectory(java.io.File,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyException(org.lambda.actions.Action0)
public static void org.approvaltests.Approvals.verifyException(org.lambda.actions.Action0,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyHtml(java.lang.String)
public static void org.approvaltests.Approvals.verifyHtml(java.lang.String,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyJson(java.lang.String)
public static void org.approvaltests.Approvals.verifyJson(java.lang.String,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyXml(java.lang.String)
public static void org.approvaltests.Approvals.verifyXml(java.lang.String,org.approvaltests.core.Options)
public static void org.approvaltests.Approvals.verifyXml(java.lang.String,org.lambda.functions.Function1)
public static void org.approvaltests.Approvals.verifyXml(java.lang.String,org.lambda.functions.Function1,org.approvaltests.core.Options)
